Replace resources TOC with in-page navigation

nick-mon1 opened the pull request:
### Summary

Replace `hugo`'s table of contents with USWDS [in-page navigation](https://designsystem.digital.gov/components/in-page-navigation/). New code removes `get_related` & `get_sharetools` to work with the new component. 

We found an issue with a11y when the user tabs through the links, they don't get to the Table of contents sidebar on the right till after they have tabbed through the content. In-page navigation addresses this but creates other layout issues listed below.

See https://github.com/uswds/uswds/issues/5136 for feature requests

### Markup Changes


**Existing**

```
<header class="usa-header usa-header--extended" role="banner">...</header>
<main role="main" id="main-content">.
  <article>Content...</article>
</main>
```

**New**

```
<main role="main" id="main-content">
  <article class="grid-container grid-container-desktop">
    <header>Header</header>
    <div class="usa-in-page-nav-container">
      <aside class="usa-in-page-nav" data-title-heading-level="h2">
        <div class="content__related-items">
          Communities & Services...
        </div>
      </aside>
      <article>
        <section>
          Content...
        </section>
      </article>
    </div>
  </article>
</main>
```

### In-page nav visibility

The in-page navigation component will display `h2` and `h3` headers as links in the sidebar but we have coded it to only show `h2`'s. This will prevent nested `h3`'s in short code components from displaying long link headers.

See this page as example — [Accordion shortcode](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/required-web-content-and-links/?dg).  It is displaying `strong` and inline text that is wrapped in an `h3`.

If there are less than 3 headers in the content, then the in-page navigation will not display.

Will display at the top of the sidebar above all other content.

### Page Examples

When navigating via the keyboard, `tabbing` down the `in-page navigation` works as expected.
Presing the `spacebar` jumps to the nav link, pressing `return/enter` will focus on the content header.

- [Good example](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/federal-web-council/?dg)

- [One item example](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/guidance-on-the-use-of-challenges-and-prizes-to-promote-open-government/?dg)

- [3 item example](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/content-management-systems-used-by-government-agencies/?dg)

- [Zero item example](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/contact-center-guidelines/?dg)

- [in-page navigation exceeds height of page](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/checklist-of-requirements-for-federal-digital-services/?dg)

- [`h3`'s are displayed causing super long navigation height](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/customer-experience-toolkit/?dg), [and here](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/lessons-from-the-2021-federal-plain-language-report-card/?dg)



**Before**
<img width="1006" alt="Screen Shot 2023-02-14 at 3 08 36 PM" src="https://user-images.githubusercontent.com/104778659/218850419-416846d6-0bbf-4d6e-8265-162821c1249c.png">

https://user-images.githubusercontent.com/104778659/218850245-d20cad0a-b686-408d-90af-1fdbecaed249.mov



**After**
<img width="981" alt="Screen Shot 2023-02-14 at 3 08 24 PM" src="https://user-images.githubusercontent.com/104778659/218850470-a17294aa-0098-40a4-8c38-6efc5a47cb94.png">

https://user-images.githubusercontent.com/104778659/218850267-af700216-dc6a-41f1-93f3-bfd7f9188c28.mov



### TODOS:
- [x] In-page keyboard accessibility works?
- [x] Hide `h3`'s from the `accordion` short code to not be displayed in the navigation?
- [x] Need to replace the border underneath the primary navigation
- [x] Audit resource pages for length requirements for in-page navigation
- [x] fix scrolling `rootMargin` to smoothly jump from link to link
- [x] fix long headlines that break container layout

mejiaj commented on themes/digital.gov/layouts/resources/single.html:
```diff
@@ -1,73 +1,82 @@
+    
+              </div>
+    
+              <!-- <div class="grid-col-12 tablet:grid-col-3">
+    
+                {{/* Table of contents */}}
+                {{- partial "core/toc.html" . -}}
+    
+                {{/* Last updated */}}
+                {{/* {{- partial "last-updated.html" . -}}  */}}
+    
+                {{/* Related Communities and Services */}}
+                {{- partial "core/get_related.html" . -}}
+    
+                {{/* Topics */}}
+                {{/* {{- partial "core/list-topics.html" . -}} */}}
+    
+                {{/* Share Tools */}}
+                {{/* {{- partial "core/get_sharetools.html" . -}} */}}
+    
+              </div> -->
```
@nick-mon1 we should try keeping these previous components. It will take some customization, but it _seems_ possible.

See https://github.com/GSA/digitalgov.gov/pull/5800/commits/9cffdb10cd08d841bee42afc5d04bc6b5eed704e

mejiaj replied:
We should also fix this extra margin on the promo wrapper

![image](https://user-images.githubusercontent.com/3385219/217663143-6e596f0d-e0ad-43cd-bfe7-505cbbeebba3.png)
https://github.com/GSA/digitalgov.gov/blob/9cffdb10cd08d841bee42afc5d04bc6b5eed704e/themes/digital.gov/src/scss/new/_promos.scss#L19

mejiaj reviewed (commented):
Some good discoveries @nick-mon1!

The [component guidance](https://designsystem.digital.gov/components/in-page-navigation/#when-to-consider-something-else-in-page-navigation) states to consider something else when dealing with short content.

Maybe we could re-work https://github.com/GSA/digitalgov.gov/blob/9cffdb10cd08d841bee42afc5d04bc6b5eed704e/themes/digital.gov/static/js/toc.js so that hides the nav on X items or less.

nick-mon1 commented:
@mejiaj Made the following updates:

- [x] removed margin-bottom for promos
- [x] in-page navigation shows first, followed by communities and services, a11y tabbing works as expected
- [x] refactor border on template instead of main
- [x] cleaned up markup, removed unused classes

mejiaj commented on themes/digital.gov/src/js/common/in-page-navigation.js:
```diff
@@ -0,0 +1,8 @@
+document.addEventListener("DOMContentLoaded", (event) => {
+  let relatedItems = document.querySelector(".content__related-items");
+  let inPageNav = document.querySelector(".usa-in-page-nav__nav");
```
Consider using `const` if these don't change.

nick-mon1 replied:
Updated with `const`.

mejiaj commented on themes/digital.gov/src/js/common/in-page-navigation.js:
```diff
@@ -0,0 +1,8 @@
+document.addEventListener("DOMContentLoaded", (event) => {
```
Event doesn't seem to be used. Can we remove?

```suggestion
document.addEventListener("DOMContentLoaded", () => {
```

nick-mon1 replied:
Good catch, this is removed.

mejiaj commented on themes/digital.gov/src/js/common/in-page-navigation.js:
```diff
@@ -0,0 +1,8 @@
+  let relatedItems = document.querySelector(".content__related-items");
+  let inPageNav = document.querySelector(".usa-in-page-nav__nav");
+
+  if (relatedItems) {
+    relatedItems.before(inPageNav);
+  }
```
Alternatively we could return early if there aren't related items.

```js
if (!relatedItems) {
  return;
}

relatedItems.before(inPageNav);
```

mejiaj replied:
Do we also want to check the length and hide `nav.usa-in-page-nav__nav` if there are less than 3?

nick-mon1 replied:
I've updated the code to hide `in-page nav` with less than 3 items. Like the simpler logic you suggested!

mejiaj commented on themes/digital.gov/src/scss/new/_in-page-navigation.scss:
```diff
@@ -0,0 +1,17 @@
+  &__item--sub-item {
+    display: none;
+  }
+}
```
Should this be in `_uswds-overrides.scss`?

nick-mon1 replied:
I see your point, moved to `_uswds-overrides.scss`.

mejiaj replied:
Let's move all the overriding styles to the overrides partial and delete this file.

mejiaj commented on themes/digital.gov/src/scss/new/_in-page-navigation.scss:
```diff
@@ -0,0 +1,17 @@
+
+  &__nav {
+    padding-left: 0;
+    padding-top: units('05');
```
Consistent quotes.

```suggestion
    padding-top: units("05");
```

mejiaj commented on themes/digital.gov/src/scss/new/_resources.scss:
```diff
@@ -1,7 +1,20 @@
 @use "uswds-core" as *;
 
+
+// Overide uswds styles for integrating in-page navigation on right column
+
+.grid-container.grid-container-desktop:has(div.usa-in-page-nav-container) {
+  // border-top: 1px solid #dfe1e2;
+}
+
+// nav.usa-nav {
+//   border-bottom: 1px solid #dfe1e2;
+// }
```
Let's remove commented code.

nick-mon1 replied:
This is now removed.

mejiaj commented on themes/digital.gov/src/scss/new/_resources.scss:
```diff
@@ -1,7 +1,20 @@
 // Styles for the main resources page
 
+.resources__container {
+  border-top: 1px solid #dfe1e2;
```
Where is this class used?

---

Consider using a USWDS theme color token like `base-lighter` (`#dfe1e2`).

mejiaj commented on themes/digital.gov/static/js/toc.js:
```diff
@@ -16,7 +16,6 @@ jQuery(function ($) {
-
 		// checks if it is a mobile browser
 		if (mobile_check() == true) {
 			var num = $('nav#TableOfContents ul:first-child > li').size();
```
Is this file still needed? I thought we were replacing this.

nick-mon1 replied:
There are other page that still use this hugo variable. This can be removed when we replace it on the other content types. Will make a note to remove this file.

mejiaj replied:
Can you create a trello card to bring in-page nav to other pages too?

mejiaj reviewed (changes requested):
@nick-mon1 component looks good and I like the simplified markup.

The sidebar disappears on mobile too. I think it improves clarity, but just wanted to call that out. I've added some small comments below, please let me know what you think. It's almost there!

nick-mon1 commented on themes/digital.gov/layouts/resources/single.html:
```diff
@@ -1,73 +1,53 @@
 
-        </div>
-      </div>
-    </section>
+              {{/* Content */}}
+              <div class="content usa-prose">
+                {{- .Content -}}
+              </div>
 
-  </article>
-</main>
-{{- end -}}
+              {{- partial "core/page-data.html" (dict "page" . "page_type" "resource" "branch" ($.Scratch.Get "branch")) -}}
+            </div>
+          </section>
+        </article>
```
Agreed, removed these tags.

mejiaj commented on themes/digital.gov/src/js/common/in-page-navigation.js:
```diff
@@ -0,0 +1,18 @@
+  }
+
+  if (navItems.length < 3) {
+    inPageNav.remove();
```
We should consider hiding via `display: none` instead of removing once we roll this out to other pages.

nick-mon1 replied:
Update the code to `inPageNav.style.display = none` so that it is ready for future changes.

mejiaj commented on themes/digital.gov/src/js/common/in-page-navigation.js:
```diff
@@ -0,0 +1,18 @@
+  const inPageNav = document.querySelector(".usa-in-page-nav__nav");
+  const navItems = document.querySelectorAll("li.usa-in-page-nav__item:not(.usa-in-page-nav__item--sub-item)");
+
+  console.log(navItems.length);
```
Let's remove this log statement.

mejiaj commented on themes/digital.gov/layouts/resources/single.html:
```diff
@@ -1,73 +1,49 @@
-
-            {{/* Last updated */}}
-            {{/* {{- partial "last-updated.html" . -}}  */}}
+      <div class="usa-in-page-nav-container">
```
The unique requirements of this template is causing sidenav headers to appear as in-page-nav items.

![image](https://user-images.githubusercontent.com/3385219/219107808-5a0bcc23-a17d-4c45-96b6-3881491078a8.png)

Right now we're hiding them, but as we're cleaning up these templates we should fix this layout issue. Meaning, we separate `<main>` from the in-page nav `<aside>`.

nick-mon1 replied:
We should discuss this, I'm curious how we can structure the markup.

mejiaj commented on themes/digital.gov/src/scss/new/_uswds-overrides.scss:
```diff
@@ -2,4 +2,8 @@
+}
+
+.usa-in-page-nav__item--sub-item {
+  @include u-display('none')
```
Let's run prettier on this file to get consistent double quotes.

nick-mon1 replied:
Done and done

nick-mon1 commented:
@ToniBonittoGSA caught this issue which relates to how the intersection observer is calculating the in view element and syncing the styling for the navigation links. Basically, when there are multiple `h2`'s within the viewport, it skips the middle ones and jumps to the intersecting one at the top. Most notably when scrolling upwards from the bottom.

This effects pages with lots of headers and small content sections, see this [page](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/checklist-of-requirements-for-federal-digital-services/?dg) 

https://user-images.githubusercontent.com/104778659/220471329-de7fc7d0-4ee7-487d-84e5-8c67987c8593.mov

nick-mon1 commented:
@ToniBonittoGSA Looks like this is a content related bug due to a long header that broke out of the container element.
I've updated the header to avoid this, but we can make some additional changes in the future to fix this on other pages.


**Breaks Layout**
[Long Header Breaks Layout: Open Government/Data/Content](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/checklist-of-requirements-for-federal-digital-services/?dg)

**Working**
[Customer Experience Toolkit](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-add-in-page-nav/resources/customer-experience-toolkit/?dg)

## Fixes

**Option 1**
Add spaces between the `/` to break the words when wrapping:

```
Open Government/Data/Content // breaks container layout
Open Government / Data / Content // wraps
```

**Option 2**
Change font-size from `1.75rem` to `1.5rem`:

```
.usa-prose>h2 {
    font-size: 1.5rem;
}
```
